### PR TITLE
Cluster managed resource

### LINKED ISSUE #7: Implement cluster managed resource
### What problem are you facing?
provider-argocd does not have support for [clusters](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/#clusters)

### How could Crossplane help solve your problem?
Implement [cluster](https://pkg.go.dev/github.com/argoproj/argo-cd/v2@v2.0.1/pkg/apis/application/v1alpha1#Cluster) resource

### PR DESCRIPTION
**Pull request by Benedikt1992**
### Description of your changes

Fixes #7

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

Tests are in progress

### How has this code been tested
Manually with the following configurations:

```yaml
---
apiVersion: cluster.argocd.crossplane.io/v1alpha1
kind: Cluster
metadata:
  name: example-cluster
spec:
  forProvider:
    server: https://kubernetes.default.svc
    name: example-cluster
    config:
      tlsClientConfig:
        insecure: true
  providerConfigRef:
    name: argocd-provider
```

```yaml
---
apiVersion: cluster.argocd.crossplane.io/v1alpha1
kind: Cluster
metadata:
  name: example-cluster
spec:
  forProvider:
    server: https://kubernetes.default.svc
    name: example-cluster
    config:
      bearerTokenRef:
        name: secret-bearer
        namespace: default
        key: password
      tlsClientConfig:
        insecure: true
  providerConfigRef:
    name: argocd-provider
```

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: apis/cluster/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	Name string `json:"name"`
+	// Config holds cluster information for connecting to a cluster
+	Config ClusterConfig `json:"config"`
+	// Holds list of namespaces which are accessible in that cluster. Cluster level resources will be ignored if namespace list is not empty.
+	// +optional
+	Namespaces []string `json:"namespaces,omitempty"`
+	// RefreshRequestedAt holds time when cluster cache refresh has been requested
+	// TODO can we remove this field?
+	// +optional
+	RefreshRequestedAt *metav1.Time `json:"refreshRequestedAt,omitempty"`
```
Comment (Benedikt1992) [inline apis/cluster/v1alpha1/types.go:39]: I feel like wee don't need this field in the managed resource. What do you think?

File: apis/cluster/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	// CertData holds PEM-encoded bytes (typically read from a client certificate file).
+	// CertData takes precedence over CertFile
+	// +optional
+	CertData []byte `json:"certData,omitempty"`
+	// +optional
+	CertDataRef *SecretReference `json:"certDataRef,omitempty"`
+	// KeyData holds PEM-encoded bytes (typically read from a client certificate key file).
+	// KeyData takes precedence over KeyFile
+	// +optional
+	KeyData []byte `json:"keyData,omitempty"`
```
Comment (Benedikt1992) [inline apis/cluster/v1alpha1/types.go:117]: To me it seems that they just used the default `TLSClientConfig` struct ([TLSClientConfig](https://pkg.go.dev/github.com/argoproj/argo-cd/v2@v2.0.1/pkg/apis/application/v1alpha1#TLSClientConfig))

It doesn't really make sense to have a Key field. Should we remove the field?

**Review by janwillies — CHANGES_REQUESTED** [review]
thanks a lot @Benedikt1992! Some small comments already in a first review

File: apis/cluster/v1alpha1/types.go (hunk 2 above)
Comment (Benedikt1992) [reply apis/cluster/v1alpha1/types.go]: We copy the argocd api

File: apis/cluster/v1alpha1/types.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+// ClusterConfig holds cluster information for connecting to a cluster
+type ClusterConfig struct {
+	// Server requires Basic authentication
+	// +optional
+	Username *string `json:"username,omitempty"`
+	// +optional
+	Password *string `json:"password,omitempty"`
+	// PasswordRef contains a reference to a kubernetes secret containing the Password
+	// +optional
+	PasswordRef *SecretReference `json:"passwordRef,omitempty"`
```
[the remark this answers is not in the document]
Comment (Benedikt1992) [reply apis/cluster/v1alpha1/types.go]: Done. This applies to Repositories as well. Will fix this in another PR

File: apis/cluster/v1alpha1/types.go [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+	Shard *int64 `json:"shard,omitempty"`
+}
+
+// ClusterConfig holds cluster information for connecting to a cluster
+type ClusterConfig struct {
+	// Server requires Basic authentication
+	// +optional
+	Username *string `json:"username,omitempty"`
+	// +optional
+	Password *string `json:"password,omitempty"`
```
[the remark this answers is not in the document]
Comment (Benedikt1992) [reply apis/cluster/v1alpha1/types.go]: Indeed it will end up in plaintext. removed original fields

**Review by janwillies — APPROVED** [review]
Thanks a lot @Benedikt1992! This is crucial for adding Kubernetes cluster to ArgoCD in a secure way via crossplane 👍

**Pull request merged**